Automattic/wp-calypso
# PoC: Try a dynamic, extensible Calypso state using wp.data

NOTICE: The lockfile is over 3000 lines of this change, please don't be scared of the line count.

NOTE: This is a proof-of-concept. It's messy and not intended to be merged in this state (pun intended!)
Also note: The line count is way inflated because of shrinkwrap updates.

This utilizes `wp.data`'s encapsulation of states into separate stores as a mechanism to break up the monolithic state we've got in Calypso. The vision is this:

* Each directory under `client/state` gets its own `wp.data` "child" store which still uses the same common Calypso redux "parent" store.
* Each child store loads its own reducer dynamically instead of all of them being imported in `client/state/index.js`.
* New code uses `withSelect` or other `wp.data` mechanisms to get selectors or dispatchable action functions.
* Existing code continues to use selectors and actions the way it does today, and can be updated to use `wp.data` when it makes sense.
* A child store is registered right before it's used, either via the `UseDataStore` component or by importing and using the selector file directly.
* All of this state code becomes much more portable to environments outside of wordpress.com (e.g. wp-admin) because of primarily living within `wp.data`.

To test:
1. `npm start`
2. Visit `http://calypso.localhost:3000/me/account` and open your dev console.

I've got a lot of temporary `console.log` lines in this PR right now to get a better understanding of the order of loading and usage of state elements. Viewing these might help to understand what's going on.

Where do we go from here?

I'd like to figure out a more central way to ensure that the child stores are loaded, I'd love to do this just right in the selectors code, so it happens on import. But the output shows that the selectors are imported even before `state/index` is, and therefore before the redux store is even created. In order to do it this way, we'd have to create the redux store and parent `wp.data` store earlier on. But if we could do this, we could avoid wrapping each selector.
